clippy_lints/src/empty_with_brackets.rs

```diff
@@ -97,22 +113,63 @@ impl EarlyLintPass for EmptyWithBrackets {
         }
     }
 
-    fn check_variant(&mut self, cx: &EarlyContext<'_>, variant: &Variant) {
+    fn check_variant(&mut self, cx: &LateContext<'_>, variant: &Variant<'_>) {
+        // Don't lint pub enums
+        if cx.effective_visibilities.is_reachable(variant.def_id) {
+            return;
+        }
+
         let span_after_ident = variant.span.with_lo(variant.ident.span.hi());
 
-        if has_brackets(&variant.data) && has_no_fields(cx, &variant.data, span_after_ident) {
+        if has_no_fields(cx, &variant.data, span_after_ident) {
+            match variant.data {
+                VariantData::Struct { .. } => {
+                    span_lint_and_then(
+                        cx,
+                        EMPTY_ENUM_VARIANTS_WITH_BRACKETS,
+                        span_after_ident,
+                        "enum variant has empty brackets",
+                        |diagnostic| {
+                            diagnostic.span_suggestion_hidden(
+                                span_after_ident,
+                                "remove the brackets",
+                                "",
+                                Applicability::MaybeIncorrect,
+                            );
+                        },
+                    );
+                },
+                VariantData::Tuple(..) => {
+                    if let Some(local_def_id) = variant.data.ctor_def_id()
+                        && !self.empty_tuple_enum_variants.contains_key(&local_def_id.to_def_id())
+                    {
+                        self.empty_tuple_enum_variants
+                            .insert(local_def_id.to_def_id(), Usage::Unused(span_after_ident));
+                    }
+                },
```

Comment:
The `.ctor_def_id()` call is redundant, the constructor `LocalDefId` is already in the `VariantData::Tuple`.

```suggestion
                VariantData::Tuple(.., local_def_id) => {
                    self.empty_tuple_enum_variants
                        .entry(local_def_id)
                        .or_insert(Usage::Unused(span_after_ident));
                },
```